internal/backend/runtime/omni/controllers/omni/machine.go

```diff
@@ -38,57 +42,116 @@ func NewMachineController() *MachineController {
+
+func (h *machineControllerHelper) transform(ctx context.Context, r controller.Reader, _ *zap.Logger, link *siderolink.Link, machine *omni.Machine) error {
+	if err := h.handleInfraProvider(ctx, r, link, machine); err != nil {
+		return err
+	}
+
+	// convert SideroLink subnet to an IP address
+	ipPrefix, err := netip.ParsePrefix(link.TypedSpec().Value.NodeSubnet)
+	if err != nil {
+		return err
+	}
+
+	helpers.CopyLabels(link, machine, omni.LabelMachineRequest, omni.LabelMachineRequestSet)
+
+	spec := machine.TypedSpec().Value
+
+	spec.ManagementAddress = ipPrefix.Addr().String()
+	spec.Connected = link.TypedSpec().Value.Connected
+
+	machine.Metadata().Labels().Set(omni.MachineAddressLabel, spec.ManagementAddress)
+
+	return nil
+}
+
+// handleInfraProvider checks if the link has an infra provider ID annotation, and if so, runs the infra provider handling logic - static or provisioning (regular).
+func (h *machineControllerHelper) handleInfraProvider(ctx context.Context, r controller.Reader, link *siderolink.Link, machine *omni.Machine) error {
+	infraProviderID, ok := link.Metadata().Annotations().Get(omni.LabelInfraProviderID)
+	if !ok { // not created by an infra provider, skip
+		return nil
+	}
+
+	providerStatus, err := safe.ReaderGetByID[*infra.ProviderStatus](ctx, r, infraProviderID)
+	if err != nil {
+		return err
+	}
+
+	if _, isStaticProvider := providerStatus.Metadata().Labels().Get(omni.LabelIsStaticInfraProvider); isStaticProvider { // static infra provider flow, e.g., bare-metal
+		return h.handleStaticInfraProvider(ctx, r, machine)
+	}
+
+	// provisioning (regular) infra provider flow, e.g., kubevirt
+	return h.handleProvisioningInfraProvider(ctx, r, link, machine)
+}
+
+// handleStaticInfraProvider checks if the machine managed by a static infra provider was accepted by the user.
+func (h *machineControllerHelper) handleStaticInfraProvider(ctx context.Context, r controller.Reader, machine *omni.Machine) error {
+	machine.Metadata().Labels().Set(omni.LabelIsManagedByStaticInfraProvider, "")
+
+	infraMachine, err := safe.ReaderGetByID[*infra.Machine](ctx, r, machine.Metadata().ID())
+	if err != nil && !state.IsNotFoundError(err) {
+		return err
+	}
+
+	accepted := infraMachine != nil && infraMachine.TypedSpec().Value.Accepted
+
+	if !accepted {
+		return xerrors.NewTaggedf[qtransform.SkipReconcileTag](
+			"the link is created by a static infra provider, but the machine is not yet accepted",
+		)
+	}
+
+	return nil
+}
+
+// handleProvisioningInfraProvider checks if the machine managed by a provisioning infra provider has the machine request and machine request set labels set.
+//
+// It then matches the machine request set owner to the machine provision controller, and if it matches, sets the no manual allocation label on the machine.
+func (h *machineControllerHelper) handleProvisioningInfraProvider(ctx context.Context, r controller.Reader, link *siderolink.Link, machine *omni.Machine) error {
```

Comment:
better naming suggestions are welcome - my thought process here was: we have 2 distinct types of infra providers: "**static**", e.g., bare-metal, and "**provisioning**", e.g. kubevirt, aws etc.